### PR TITLE
Attempting fix for flakey notifier test

### PR DESCRIPTION
**Pull request by AlexHentschel**
* added readme explaining usage of notifier; 
* revisions to notifier's unit test `TestNotifier_AllWorkProcessed` to hopefully better deal with overwhelmed testing system

### CODE REVIEW

File: engine/notifier_test.go [hunk 1]
```diff
@@ -83,54 +83,62 @@ func TestNotifier_ManyNotifications(t *testing.T) {
...
+		startingWorkers.Wait()
 
-	// send 100 notifications, with small delays
-	for i := 0; i < 100; i++ {
-		notifier.Notify()
-		time.Sleep(10 * time.Microsecond)
+		// send 100 notifications, with small delays
+		for i := 0; i < 100; i++ {
+			notifier.Notify()
+		}
```
Comment (smnzhu) [inline engine/notifier_test.go:107]: This is probably unlikely, but I do see a race condition here *in theory*. Let's reduce the example to 2 workers instead of 100 for simplicity. 

Consider the following execution order for the 3 threads (the main thread, and the 2 worker routines):
* worker 1 executes line 97
* worker 2 executes line 97
* main thread returns from call on line 102
* main thread executes line 105 (first loop iteration)
* main thread executes line 106 (first loop iteration)
* main thread executes line 105 (second loop iteration)
* main thread executes line 106 (second loop iteration) *NOTIFICATION DROPPED*
* worker 1 executes line 98 (returns immediately)
* worker 2 executes line 98 (blocks forever)

In short, the loop in the main thread sends *both* notifications before either worker routine has a chance to read from the notification channel, which results in the second routine never being notified.

All this to say, perhaps it would be good to keep the `time.Sleep()` after each loop iteration?

File: engine/notifier_test.go [hunk 2]
```diff
@@ -141,24 +149,25 @@ func TestNotifier_AllWorkProcessed(t *testing.T) {
...
 				for consumedWork.Load() < totalWork {
 					<-notifier.Channel()
-				L:
 					for {
 						select {
 						case <-pendingWorkQueue:
 							consumedWork.Inc()
 						default:
-							break L
+							break
```
Comment (smnzhu) [inline engine/notifier_test.go:163]: Did we not want to keep the label [as discussed here](https://github.com/onflow/flow-go/pull/1037#issuecomment-887896284)?

**Comment by smnzhu** [comment]
Thanks @AlexHentschel! [This](https://github.com/onflow/flow-go/pull/1037#issuecomment-888017485) explanation makes a lot of sense, and your test results do seem to indicate that this will hopefully fix the issue. Just had a few comments :)

File: engine/notifier_test.go (hunk 1 above)
Comment (AlexHentschel) [reply engine/notifier_test.go:107]: great catch. totally agree.

File: engine/notifier_test.go (hunk 2 above)
Comment (AlexHentschel) [reply engine/notifier_test.go]: thanks for the reminder. Absolutely this should break the `for` loop
